cpp/src/copying/copy_range.cuh

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bitmask/bit_mask.cuh>
+#include <utilities/error_utils.hpp>
+#include <utilities/type_dispatcher.hpp>
+#include <utilities/bit_util.cuh>
+#include <utilities/cuda_utils.hpp>
+#include <utilities/column_utils.hpp>
+#include <string/nvcategory_util.hpp>
+
+#include <cub/cub.cuh>
+
+namespace {
+
+using bit_mask::bit_mask_t;
+static constexpr gdf_size_type warp_size{32};
+
+template <typename T, typename InputFunctor, bool has_validity>
+__global__
+void copy_range_kernel(T * __restrict__ const data,
+                       bit_mask_t * __restrict__ const bitmask,
+                       gdf_size_type * __restrict__ const null_count,
+                       gdf_index_type begin,
+                       gdf_index_type end,
+                       InputFunctor input)
+{
+  const gdf_index_type tid = threadIdx.x + blockIdx.x * blockDim.x;
+  constexpr size_t mask_size = warp_size;
+
+  const gdf_size_type masks_per_grid = gridDim.x * blockDim.x / mask_size;
+  const int warp_id = tid / warp_size;
+  const int lane_id = threadIdx.x % warp_size;
+
+  const gdf_index_type begin_mask_idx =
+      cudf::util::detail::bit_container_index<bit_mask_t>(begin);
+  const gdf_index_type end_mask_idx =
+      cudf::util::detail::bit_container_index<bit_mask_t>(end);
+
+  gdf_index_type mask_idx = begin_mask_idx + warp_id;
+
+  gdf_index_type output_offset = begin_mask_idx * mask_size - begin;
+  gdf_index_type input_idx = tid + output_offset; 
+
+  // each warp shares its total change in null count to shared memory to ease
+  // computing the total change to null_count.
+  // note maximum block size is limited to 1024 by this, but that's OK
+  __shared__ uint32_t warp_null_change[has_validity ? warp_size : 1];
+  if (has_validity && threadIdx.x < warp_size) warp_null_change[threadIdx.x] = 0;
+
+  __syncthreads(); // wait for shared data and validity mask to be complete
+
+  while (mask_idx <= end_mask_idx)
+  {
+    gdf_index_type index = mask_idx * mask_size + lane_id;
+    bool in_range = (index >= begin && index < end);
+
+    // write data
+    if (in_range) data[index] = input.data(input_idx);
+
+    if (has_validity) { // update bitmask
+      int active_mask = __ballot_sync(0xFFFFFFFF, in_range);
+
+      bool valid = in_range & input.valid(input_idx);
```

Comment:
This was doing a bit-wise `AND`, which is technically correct for booleans here, but I think it's safer/clearer to just use a logical `AND`. 
```suggestion
      bool valid = in_range and input.valid(input_idx);
```

Reply:
Yikes, good catch, I meant `&&`.